### PR TITLE
Fix data race on TrainingLinearLayer's transposed weight, and make test tokenisation deterministic

### PR DESCRIPTION
**Pull request by MBrekhof**
Two causes of non-deterministic test results, one a real product bug.

## 1. Data race on `TrainingLinearLayer`'s transposed weight

`FfnKernels.MoE` runs `ProcessToken` over tokens with `Parallel.For`, and every token calls `Forward` on the **same** expert `LinearLayer` instances. `TrainingLinearLayer.MatMulForward` did:

```csharp
_weightBT?.Dispose();
_weightBT = _weight.Transpose();
fn(input.DataPtr, (byte*)_weightBT.DataPtr, output.DataPtr, ...);
```

`_weightBT` is shared instance state, so one thread disposed (and returned to `NativeBufferPool`) the transposed weight another thread was mid-matmul on, and both raced to publish the field. The result was silently wrong logits that differed run to run.

This is what made `SmmExportLoadTests.TrainingExport_MoE_ReloadsWithParity` flaky: run alone it failed 4 times in 5, with the reloaded model's logits varying between runs while the original's stayed fixed. Forcing the MoE token loop serial made it pass 5/5, which pinned it to the parallel path.

`_weightBT` was never a cache — it was disposed and recomputed on every call, and nothing outside the two methods that write it ever read it. Making it a local `using` removes the shared state entirely, so no locking is needed and the buffer is freed at end of call instead of lingering. `InvalidateCache` had nothing left to do; the base is a no-op.

`InferenceLinearLayer` writes no instance state in `Forward`, so the inference path was never affected — training only.

## 2. Test token mapping was randomised per process

Five test files mapped words to token ids with `Math.Abs(w.GetHashCode()) % vocabSize`. String hashing is seeded per process in .NET, so every `dotnet test` invocation produced a different corpus→token mapping. Verified directly — the same seven words hashed to three completely different id sequences in three consecutive processes.

For most of those tests the ids are opaque, but `QuantAwareTrainingTests.TrainLoop_WithQat_LossDescends` asserts a numeric training outcome (`losses[^1] < losses[0]` after 12 steps), so it trained on effectively random data and passed or failed by luck — roughly 2 runs in 6. It also explains why all five theory variants failed in lockstep: they share one process, so one unlucky seed sank all of them.

Replaced with a shared FNV-1a helper in `TestHelpers`.

## Verification

- MoE parity test: 10/10 alone (was 4 failures in 5); full suite 644/644 across 8 consecutive runs, where `master` gave 3 failures in 6.
- QAT theory: 6/6 with all 5 variants.

### CODE REVIEW

**Comment by Integral2u** [comment]
Reviewed and merged into `master` (fast-forward). The PR content is live in commits: 22ad646, 0200b33. Closing this PR.
